kyuubi-server/src/main/scala/org/apache/kyuubi/engine/trino/TrinoProcessBuilder.scala

```diff
@@ -0,0 +1,153 @@
+          .resolve(module)
+          .toFile)
+        .map(_.getAbsolutePath)
+    }
+
+    trinoHomeOpt.map { dir =>
+      Paths.get(dir, "bin", TRINO_ENGINE_BINARY_FILE).toAbsolutePath.toFile.getCanonicalPath
+    }.getOrElse {
+      throw KyuubiSQLException("TRINO_ENGINE_HOME is not set! " +
+        "For more detail information on installing and configuring Trino, please visit " +
+        "https://kyuubi.apache.org/docs/stable/deployment/settings.html#environments")
+    }
+  }
+
+  override protected def mainResource: Option[String] = {
+    val jarName = s"${module}_$SCALA_COMPILE_VERSION-$KYUUBI_VERSION.jar"
+    // 1. get the main resource jar for user specified config first
+    conf.get(ENGINE_TRINO_MAIN_RESOURCE).filter { userSpecified =>
+      // skip check exist if not local file.
+      val uri = new URI(userSpecified)
+      val schema = if (uri.getScheme != null) uri.getScheme else "file"
+      schema match {
+        case "file" => Files.exists(Paths.get(userSpecified))
+        case _ => true
+      }
+    }.orElse {
+      // 2. get the main resource jar from system build default
+      env.get(KyuubiConf.KYUUBI_HOME)
+        .map { Paths.get(_, "externals", "engines", "trino", "jars", jarName) }
+        .filter(Files.exists(_)).map(_.toAbsolutePath.toFile.getCanonicalPath)
+    }.orElse {
+      // 3. get the main resource from dev environment
+      Option(Paths.get("externals", module, "target", jarName))
+        .filter(Files.exists(_)).orElse {
+          Some(Paths.get("..", "externals", module, "target", jarName))
+        }.map(_.toAbsolutePath.toFile.getCanonicalPath)
+    }
+  }
+
+  override protected def module: String = "kyuubi-trino-engine"
+
+  override protected def mainClass: String = "org.apache.kyuubi.engine.trino.TrinoSqlEngine"
+
+  override protected def childProcEnv: Map[String, String] = conf.getEnvs +
+    ("TRINO_ENGINE_JAR" -> mainResource.get) +
+    ("TRINO_ENGINE_DYNAMIC_ARGS" ->
+      trinoConf.map { case (k, v) => s"-D$k=$v" }.mkString(" "))
+
+  override protected def commands: Array[String] = Array(executable)
+
+  override protected val workingDir: Path = {
+    env.get("KYUUBI_WORK_DIR_ROOT").map { root =>
+      val workingRoot = Paths.get(root).toAbsolutePath
+      if (!Files.exists(workingRoot)) {
+        debug(s"Creating KYUUBI_WORK_DIR_ROOT at $workingRoot")
+        Files.createDirectories(workingRoot)
+      }
+      if (Files.isDirectory(workingRoot)) {
+        workingRoot.toString
+      } else null
+    }.map { rootAbs =>
+      val working = Paths.get(rootAbs, proxyUser)
+      if (!Files.exists(working)) {
+        debug(s"Creating $proxyUser's working directory at $working")
+        Files.createDirectories(working)
+      }
+      if (Files.isDirectory(working)) {
+        working
+      } else {
+        Utils.createTempDir(rootAbs, proxyUser)
+      }
+    }.getOrElse {
+      Utils.createTempDir(namePrefix = proxyUser)
+    }
+  }
```

Comment:
We can reuse the ProcBuilder#workingDir after merging #1758 . WDYT?

Reply:
+1

Reply:
+1, I will address it.